src/callback.c

```diff
@@ -0,0 +1,140 @@
+}
+
+static void initialize_globals(SEXP bits_sexp, SEXP seed_sexp, SEXP ntuple_sexp, 
+			       SEXP quiet_sexp, SEXP verbose_sexp) {
+
+    // FUTURE: add pmultiplier to user supplied?
+    quiet = INTEGER_VALUE(quiet_sexp);
+    if (! quiet) hist_flag = 1; // print histogram unless quiet
+    verbose = INTEGER_VALUE(verbose_sexp);
+    ntuple = INTEGER_VALUE(ntuple_sexp);
+
+    uli_t bits = INTEGER_VALUE(bits_sexp);
+    uli_t max = (1LL << bits) - 1;
+
+    gsl_rng_type callback_rng = {
+	"callback",         // arbitrary name
+	max,                // rand max (pow2 - 1)
+	0,  	            // rand min (must be 0) 
+	0,                  // no storage space
+	&callback_set_seed, // set seed 
+	&callback_get_int,  // random int
+	NULL                // no double
+    };
+
+    gsl_rng_default_seed = INTEGER_VALUE(seed_sexp); 
+    rng = gsl_rng_alloc(&callback_rng); // uses gsl_rng_default_seed
+
+    rmax_mask = max;
+    rmax_bits = bits;
+
+    overlap = 1;       /* use overlapping samples */
+    multiply_p = 1;    /* use default number of psamples */
+    gnumbs[0] = 13;    /* mt19937 as a "good" generator */
+
+    Xtrategy = 1;          /* 0 means "no strategy, off", the default */
+    Xweak = 0.005;         /* Point where generator is flagged as "weak" */
+    Xfail = 0.000001;      /* Point where generator unambiguously fails dieharder */
+    Xstep = 100;           /* Number of pvalues to add when iterating in TTD/RA */
+    Xoff = 100000;         /* This is quite stressful and may run a long time */
+}
+
+static SEXP parse_and_test_callback(SEXP callback) {
+    ParseStatus parseStatus;
+    SEXP parsedVector = R_ParseVector(callback, 1, &parseStatus, R_NilValue);
+    if (parseStatus != PARSE_OK) {
+        error("Failed to parse \'%s\' as string", CSTRING(callback));
+	return NULL;
+    }
+    
+    SEXP parsedCallback = VECTOR_ELT(parsedVector, 0);
+
+    int evalError;
+    SEXP result = R_tryEval(parsedCallback, R_GlobalEnv, &evalError);
+    if (evalError) {
+        error("Failed to eval parsed callback '%s'", CSTRING(callback));
+	return NULL;
+    }
+
+    if (TYPEOF(result) != INTSXP && TYPEOF(result) != REALSXP) {
+        error("Result from eval parsed callback '%s' is not integer or real", CSTRING(callback));
+	return NULL;
+    }
+	
+    if (length(result) != 1) {
+        error("Result from eval parsed callback '%s' is not length 1", CSTRING(callback));
+	return NULL;
+    }
+
+    if (INTEGER_VALUE(result) == NA_INTEGER)  {
+	// this might be a rare random error but safest to force retry
+	error("Result from eval parsed callback '%s' was NA", CSTRING(callback));
+        return NULL;
+    }
+    
+    if (verbose) Rprintf("Test random was: %u\n", INTEGER_VALUE(result));
+    return parsedCallback;
+}
+
+SEXP dieharderCallback(SEXP callback_sexp, SEXP bits_sexp, SEXP testnum_sexp, 
+		       SEXP seed_sexp, SEXP psamples_sexp, SEXP tsamples_sexp,
+		       SEXP ntuple_sexp, SEXP quiet_sexp, SEXP verbose_sexp) {
```

Comment:
On Wed, Sep 23, 2015 at 6:35 AM, Dirk Eddelbuettel
 >It is really high time I convert RDieHarder to Rcpp

My suggestion probably would be that you ignore these patches until until
you make the switch, and then add a more succinct callback version to the
new code using Rcpp.   The version I suggested here might be a good
template for approach.

That said, code quality of libdieharder is low enough that I don't know if
a new version is merited.  It may be better to let it fade off the map.
While the Rcpp version could be much cleaner, the problem is more with the
base library than with the wrapper.

If you were to patch the current, you'd probably want to concentrate on
getting rid of the segfault.  Yanking support for user-defined generators
would be easiest.  I'd guess that no one has ever used one, since my
attempts to do so resulted in immediate crashes.

--nate